### PR TITLE
Fixed #5362, #4029: fixed terminal on-restart leaking issue after using vim

### LINKED ISSUE #5362: [terminal] `vi` causes device status code printed out on terminal restart
This is an investigation of the old issue #4029. After testing, I found out the garbage characters printed out were actually device status code emitted by Xterm.

In `xterm/src/InputHandler`:
```ts
  public sendDeviceAttributes(params: number[], collect?: string): void {
  ...
      if (this._terminal.is('xterm')) {
        this._terminal.handler(C0.ESC + '[>0;276;0c');
      } else if (this._terminal.is('rxvt-unicode')) {
        this._terminal.handler(C0.ESC + '[>85;95;0c');
  ...
```

To reproduce the issue: 
1. open `vi` in Theia terminal, then exit
2. refresh the browser
3. as Theia restarts, the terminal restores and the device status code is printed out

### PR DESCRIPTION
**Pull request by fangnx**
Fixed #5362, #4029.

Fixed the terminal device status code leaking issue (as described in #5362) when terminal is restored.  Inside the `terminal-widget-impl.ts`, added a conditional branch that checked whether the data sent is the *device status code* emitted by Xterm. If it is *device status code*, it would be blocked from being sent through the websocket, so that it would be restored and displayed when Theia terminal restarts.

Since most of the "write" and data processing functionalities are implemented in Xterm, at this moment there may not be a better way to prevent the *device status code* (which is emitted by Xterm.js when vim is opened) from being sent to the buffer.

### CODE REVIEW

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 1]
```diff
@@ -374,6 +374,9 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
     }
 
+    // Device status code emitted by Xterm.js, in InputHandler.ts (line 1028)
```
[the remark this answers is not in the document]
Comment (fangnx) [reply packages/terminal/src/browser/terminal-widget-impl.ts]: It is in `sendDeviceAttributes()`, line 1107. I would update the comment by adding the function name ;)

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 2]
```diff
@@ -387,7 +390,13 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
             onConnection: connection => {
                 connection.onNotification('onData', (data: string) => this.write(data));
 
-                const sendData = (data?: string) => data && connection.sendRequest('write', data);
+                // Excludes the device status code emitted by Xterm.js
+                const sendData = (data?: string) => {
+                    if (data && this.deviceStatusCodes.indexOf(data) === -1) {
```
[the remark this answers is not in the document]
Comment (fangnx) [reply packages/terminal/src/browser/terminal-widget-impl.ts]: I think `Array.includes(data)` is not supported in Theia (as it is introduced in ES7) ;)

[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/terminal/src/browser/terminal-widget-impl.ts]:
```ts
const set = new Set(['a', 'b']);
console.log(set.has('a'));
```

File: packages/terminal/src/browser/terminal-widget-impl.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/terminal/src/browser/terminal-widget-impl.ts]: maybe put the github link like I did, but pointing to the right resource then :)

**Comment by fangnx** [comment]
> It blocks everything i try to send to the terminal. Also generally the approach looks hacky.
> Do you know what those codes say and whether we not want to receive them?

Hi Sven,

Just re-tested and updated the code. There was a bug in the logic that caused the blocking, which has just been fixed. Sorry for the confusion.

The codes that leaked and should be blocked are the *Device Attributes*, which is emitted by Xterm.js every time `vi/vim` is run. Applications such as `vi` request the code to get attributes of the terminal (such as the terminal type). 

When `vi` is run, the *Device Attributes* were emitted and **automatically sent to the buffer** that would be restored on terminal restart. That was the bug that caused the terminal leaking issue. Blocking the codes by a conditional branch does look hacky ;), but I don't think there is some other way to fix it on our side (without modifying Xterm.js).

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 3]
```diff
@@ -374,6 +374,10 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
     }
 
+    // Device status code emitted by Xterm.js
+    // Check: https://github.com/xtermjs/xterm.js/blob/master/src/InputHandler.ts#L1107
```
Comment (marechal-p) [inline packages/terminal/src/browser/terminal-widget-impl.ts]: Detail, but links pointing to a file on master are likely to break. Specific commits, or even release branches are better. E.g. https://github.com/xtermjs/xterm.js/blob/release/3.14/src/InputHandler.ts#L1055-L1082

**Comment by marechal-p** [comment]
I want to have a better look before merging, will try later this evening.

**Comment by marechal-p** [comment]
What is happening without the patch:
- We open a terminal in Theia, new ring buffer gets created.
- We open `vi`, send a bunch of control characters, one of which causes Xterm.js to emit a response, sent to the terminal process (`vi` here), and it gets consumed, `vi` emits some other formatting codes and keeps doing whatever it was doing, based on the received response.
- We reload the Theia window.
- The whole ring buffer is sent again. In that buffer, the character written by `vi` that makes Xterm.js send the codes is parsed again.
- Xterm.js sends the response code because it saw it in the stream, except that `vi` was closed, the running process is just the command line. Upon reception it interprets the data as regular user input.

TL;DR: The issue happens on reload because we send the whole history of outputted bytes, among which there is a request character making Xterm.js emit some response, except that the command line sees it as regular input.

What we tried:
- Add a new `onInitialized` notification, sent after the first `onData` notification.
- Until we receive the `onInitialized` message, we suppress everything Xterm.js tries to send (because it is re-parsing a past buffer).

Problem:
- Xterm.js doesn't answer in a determinist way, sometimes it is quick, sometimes it is not, sometimes it doesn't need to send anything. How should we know when to stop waiting for it?

We are out of ideas at the moment.

**Comment by fangnx** [comment]
@marechal-p I updated the commit message to explain what this patch is about.

**Comment by lmcbout** [comment]
@fangnx  @marechal-p 
I found some explanation about the escapes sequences and it says:
"I've tidied up the escape sequence expression to follow the Wikipedia overview of ANSI escape codes, focusing on the CSI sequences, and ignoring the C1 codes as they are never used in today's UTF-8 world)"

You can also find some explanation on the following site under the platform support:
https://en.wikipedia.org/wiki/ANSI_escape_code

ANSI escape sequences are a standard for in-band signaling to control the cursor location, color, and other options on video text terminals and terminal emulators. Certain sequences of bytes, most starting with Esc and '[', are embedded into the text, which the terminal looks for and interprets as commands, not as character codes.

**Comment by marechal-p** [comment]
@lmcbout yes, but in this case it is a bit different, escape sequences meant to move the cursor/change the color are sent from process to terminal. In this case, it is something that the terminal sends to the process after a certain request. And it is really hard to find documentation for this one.

**Comment by marechal-p** [comment]
I've found the following resources, but I don't understand how something like `\u001B[>0;276;0c` fits:

- https://vt100.net/docs/vt510-rm/DA1.html
- https://vt100.net/docs/vt510-rm/DA2.html
- https://vt100.net/docs/vt510-rm/DA3.html
